### PR TITLE
Expose GRPC KeepAlive parameters to triton server

### PR DESCRIPTION
**Pull request by rmccorm4**
- [x] Add GRPC KeepAlive parameters to server
- [x] Add GRPC KeepAlive parameters to `tritonserver` cli
- [x] Fix `grpc_comression_level` -> `grpc_compression_level` typo
- [x] Add GRPC KeepAlive parameters to --log-verbose output

### CODE REVIEW

File: src/servers/main.cc [hunk 1]
```diff
@@ -94,6 +94,12 @@ bool grpc_use_ssl_ = false;
 nvidia::inferenceserver::SslOptions grpc_ssl_options_;
 grpc_compression_level grpc_response_compression_level_ =
     GRPC_COMPRESS_LEVEL_NONE;
+int32_t grpc_arg_keepalive_time_ms_ = 7200000;
```
Comment (tanmayv25) [inline src/servers/main.cc:97]: Can you add a comment here saying the defaults were picked from https://grpc.github.io/grpc/cpp/md_doc_keepalive.html ?

File: src/servers/main.cc [hunk 2]
```diff
@@ -350,6 +362,37 @@ std::vector<Option> options_
        "The compression level to be used while returning the infer response to "
        "the peer. Allowed values are none, low, medium and high. By default, "
        "compression level is selected as none."},
+      {OPTION_GRPC_ARG_KEEPALIVE_TIME_MS, "grpc-keepalive-time", Option::ArgInt,
+       "The period (in milliseconds) after which a keepalive ping is sent on "
+       "the transport."},
```
Comment (tanmayv25) [inline src/servers/main.cc:367]: Add the default values for all these options in these option description as well.

**Review by tanmayv25 — CHANGES_REQUESTED** [review]

**Comment by rmccorm4** [comment]
New `tritonserver` parameters and help string:
```
root@0740505e5dd7:/opt/tritonserver# tritonserver --help                                
...
  --grpc-keepalive-time <integer>
        The period (in milliseconds) after which a keepalive ping is
        sent on the transport. Default is 7200000 (2 hours).
  --grpc-keepalive-timeout <integer>
        The period (in milliseconds) the sender of the keepalive
        ping waits for an acknowledgement. If it does not receive an
        acknowledgment within this time, it will close the connection. Default is
        20000 (20 seconds).
  --grpc-keepalive-permit-without-calls <boolean>
        Allows keepalive pings to be sent even if there are no calls
        in flight (0 : false; 1 : true). Default is 0 (false).
  --grpc-http2-max-pings-without-data <integer>
        The maximum number of pings that can be sent when there is
        no data/header frame to be sent. gRPC Core will not continue sending
        pings if we run over the limit. Setting it to 0 allows sending pings
        without such a restriction. Default is 2.
  --grpc-http2-min-recv-ping-interval-without-data <integer>
        If there are no data/header frames being sent on the
        transport, this channel argument on the server side controls the minimum
        time (in milliseconds) that gRPC Core would expect between receiving
        successive pings. If the time between successive pings is less than
        this time, then the ping will be considered a bad ping from the peer.
        Such a ping counts as a ‘ping strike’. On the client side, this
        does not have any effect. Default is 300000 (5 minutes).
  --grpc-http2-max-ping-strikes <integer>
        Maximum number of bad pings that the server will tolerate
        before sending an HTTP2 GOAWAY frame and closing the transport.
        Setting it to 0 allows the server to accept any number of bad pings.
        Default is 2.
```

Logging server-side KeepAlive Parameters when verbose mode set:
```
root@0740505e5dd7:/opt/tritonserver# tritonserver --model-store ./models --log-verbose 1 \
  --grpc-keepalive-time 1 \
  --grpc-keepalive-timeout 2 \
  --grpc-keepalive-permit-without-calls true \
  --grpc-http2-max-pings-without-data 4 \
  --grpc-http2-min-recv-ping-interval-without-data 5 \
  --grpc-http2-max-ping-strikes 6
...
I0713 23:12:28.448441 204 grpc_server.cc:4078] === GRPC KeepAlive Parameters ===
I0713 23:12:28.448449 204 grpc_server.cc:4079] grpc_arg_keepalive_time_ms_: 1
I0713 23:12:28.448455 204 grpc_server.cc:4081] grpc_arg_keepalive_timeout_ms_: 2
I0713 23:12:28.448461 204 grpc_server.cc:4083] grpc_arg_keepalive_permit_without_calls_: 1
I0713 23:12:28.448468 204 grpc_server.cc:4085] grpc_arg_http2_max_pings_without_data_: 4
I0713 23:12:28.448474 204 grpc_server.cc:4087] grpc_arg_http2_min_recv_ping_interval_without_data_ms_: 5
I0713 23:12:28.448481 204 grpc_server.cc:4089] grpc_arg_http2_max_ping_strikes_: 6
I0713 23:12:28.448489 204 grpc_server.cc:4091] ==============================
...
I0713 23:12:28.464502 204 grpc_server.cc:4124] Started GRPCInferenceService at 0.0.0.0:8001
I0713 23:12:28.465076 204 http_server.cc:2795] Started HTTPService at 0.0.0.0:8000
I0713 23:12:28.507275 204 http_server.cc:162] Started Metrics Service at 0.0.0.0:8002
```

Defaults per https://grpc.github.io/grpc/cpp/md_doc_keepalive.html:
```
root@0740505e5dd7:/opt/tritonserver# tritonserver --model-store ./models --log-verbose 1                                                                                                                                            
...
I0713 23:15:38.867531 225 grpc_server.cc:4078] === GRPC KeepAlive Parameters ===
I0713 23:15:38.867538 225 grpc_server.cc:4079] grpc_arg_keepalive_time_ms_: 7200000
I0713 23:15:38.867543 225 grpc_server.cc:4081] grpc_arg_keepalive_timeout_ms_: 20000
I0713 23:15:38.867552 225 grpc_server.cc:4083] grpc_arg_keepalive_permit_without_calls_: 0
I0713 23:15:38.867561 225 grpc_server.cc:4085] grpc_arg_http2_max_pings_without_data_: 2
I0713 23:15:38.867567 225 grpc_server.cc:4087] grpc_arg_http2_min_recv_ping_interval_without_data_ms_: 300000
I0713 23:15:38.867574 225 grpc_server.cc:4089] grpc_arg_http2_max_ping_strikes_: 2
I0713 23:15:38.867583 225 grpc_server.cc:4091] ==============================
...
I0713 23:15:38.868949 225 grpc_server.cc:4124] Started GRPCInferenceService at 0.0.0.0:8001
I0713 23:15:38.869510 225 http_server.cc:2795] Started HTTPService at 0.0.0.0:8000
I0713 23:15:38.911639 225 http_server.cc:162] Started Metrics Service at 0.0.0.0:8002
```

**Comment by rmccorm4** [comment]
@tanmayv25 made the suggested changes and a couple more if you'd like to take another look.

**Comment by tanmayv25** [comment]
Why not mark the PR Ready for review ?

**Review by tanmayv25 — CHANGES_REQUESTED** [review]

File: src/servers/main.cc [hunk 3]
```diff
@@ -350,6 +363,39 @@ std::vector<Option> options_
...
+       "pings if we run over the limit. Setting it to 0 allows sending pings "
+       "without such a restriction. Default is 2."},
+      {OPTION_GRPC_ARG_HTTP2_MIN_RECV_PING_INTERVAL_WITHOUT_DATA_MS,
+       "grpc-http2-min-recv-ping-interval-without-data", Option::ArgInt,
+       "If there are no data/header frames being sent on the transport, this "
+       "channel argument on the server side controls the minimum time "
+       "(in milliseconds) that gRPC Core would expect between receiving "
+       "successive pings. If the time between successive pings is less than "
+       "this time, then the ping will be considered a bad ping from the peer. "
+       "Such a ping counts as a ‘ping strike’. On the client side, this does "
```
Comment (tanmayv25) [inline src/servers/main.cc:392]: No need to document client side behavior of these parameters here. Just include server behavior.
Make sure you document these parameters correctly in your client side PR. 
Anyways, you would not have this option on client side as it does not have any effect.

File: src/servers/grpc_server.h [hunk 4]
```diff
@@ -81,7 +87,13 @@ class GRPCServer {
       const std::shared_ptr<SharedMemoryManager>& shm_manager,
       const std::string& server_addr, bool use_ssl,
       const SslOptions& ssl_options, const int infer_allocation_pool_size,
-      grpc_compression_level compression_level);
+      grpc_compression_level compression_level,
+      const int grpc_arg_keepalive_time_ms = 7200000,
```
[the remark this answers is not in the document]
Comment (rmccorm4) [reply src/servers/grpc_server.h:91]: Are you saying to not include default values anywhere in `grpc_server.h` or `grpc_server.cc` and make these parameters required?

I could see that requiring no user changes when using `tritonserver` CLI if we provide the default values in `main.cc`, but for users that use the library directly with C API, wouldn't that break their code moving forward?

File: src/servers/grpc_server.h [hunk 5]
```diff
@@ -93,6 +105,15 @@ class GRPCServer {
   const int infer_allocation_pool_size_;
   grpc_compression_level compression_level_;
 
+  // GRPC KeepAlive Settings:
+  // https://grpc.github.io/grpc/cpp/md_doc_keepalive.html
+  const int grpc_arg_keepalive_time_ms_;
```
[the remark this answers is not in the document]
Comment (rmccorm4) [reply src/servers/grpc_server.h:110]: Sure, I can go ahead with a struct to encapsulate these parameters

File: src/servers/grpc_server.h (hunk 4 above)
[the remark this answers is not in the document]
Comment (tanmayv25) [reply src/servers/grpc_server.h]: > Are you saying to not include default values anywhere in grpc_server.h or grpc_server.cc and make these parameters required in both Create() and in GRPCServer() constructor?

Yes

> but for users that use the library directly with C API, wouldn't that break their code moving forward?

The C API triton library does not expose grpc or http endpoints to the users.

The hierarchy of code organization is:

GRPC/HTTP endpoint -> GRPC/HTTP Server -> Triton C API -> Triton Core

The users of Triton C API will directly communicate with the Triton Core using the API and will not use GRPC/HTTP protocol.

See [here](https://github.com/triton-inference-server/server/blob/main/docs/images/arch.jpg).

**Review by tanmayv25 — APPROVED** [review]

**Pull request merged**